pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,227 @@
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// Reconcile responds to Machine events to assign a NodeRef.
+// +kubebuilder:rbac:groups=,resources=secrets,verbs=get;list;watch
+func (r *ReconcileNodeRef) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	klog.Infof("Reconcile request for Machine %q in namespace %q", request.Name, request.Namespace)
+	ctx := context.Background()
+
+	// Fetch the Machine instance.
+	machine := &v1alpha1.Machine{}
+	err := r.Get(ctx, request.NamespacedName, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).Infof("Machine %q in namespace %q is not found, won't reconcile", machine.Name, machine.Namespace)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		klog.V(2).Infof("Machine %q in namespace %q has been deleted, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine doesn't already have a NodeRef.
+	if machine.Status.NodeRef != nil {
+		klog.V(2).Infof("Machine %q in namespace %q already has a NodeRef, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a cluster label.
+	if machine.Labels[v1alpha1.MachineClusterLabelName] == "" {
+		klog.V(2).Infof("Machine %q in namespace %q doesn't specify %q label, won't reconcile", machine.Name, machine.Namespace,
+			v1alpha1.MachineClusterLabelName)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a valid Cluster associated with it.
+	cluster, err := r.getCluster(ctx, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Warningf("Cannot find a Cluster for Machine %q in namespace %q, won't reconcile", machine.Name, machine.Namespace)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, retrying later", machine.Name, machine.Namespace)
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	result, err := r.reconcile(ctx, cluster, machine)
+	if err != nil {
+		klog.Errorf("Failed to assign NodeRef to Machine %q: %v", request.NamespacedName, err)
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return result, err
+	}
+
+	klog.Infof("Set Machine's (%q in namespace %q) NodeRef to %q", machine.Name, machine.Namespace, machine.Status.NodeRef.Name)
+	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	return result, nil
+}
+
+func (r ReconcileNodeRef) reconcile(ctx context.Context, cluster *v1alpha1.Cluster, machine *v1alpha1.Machine) (reconcile.Result, error) {
```

Comment:
The other `ReconcileNodeRef` methods have pointer receivers but this one does not?